nixos/mosquitto: fix escaping, types, allow multiple listeners

pennae opened the pull request:
###### Motivation for this change

the mosquitto module interpolates stuff into shell scripts without escaping, which is a bit sad when you try to configure a literal password that contains a `'`. also some types are weird, and there's no overwhelming need to restrict the module to only one listener.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nagy commented on nixos/modules/services/networking/mosquitto.nix:
```diff
@@ -35,6 +35,29 @@ let
+      port = mkOption {
+        default = port;
+        example = port;
+        type = types.int;
```
```suggestion
        type = types.port;
```
?

pennae commented:
maybe the module would benefit from being rewritten completely? there's a lot it currently cannot do (like enabling client cert authentication on only one listener), and the password file generation is pretty wonky one way or other. we don't really validate the config either, even though we could.

pennae commented:
took a good long look at rewriting the module. due to deficiencies in the mosquitto config parser we'd have to do *a lot* of validation in the module, requiring one an option definition for almost every option mosquitto has. if that sounds like a good idea we can do that, but honestly we'd rather not.

peterhoeg commented:
> took a good long look at rewriting the module. due to deficiencies in the mosquitto config parser we'd have to do a lot of validation in the module

I'm in favour on the minimum amount of work required to get it going.

pennae committed:
nixos/mosquitto: rewrite the module

mosquitto needs a lot of attention concerning its config because it doesn't
parse it very well, often ignoring trailing parts of lines, duplicated config
keys, or just looking back way further in the file to associated config keys
with previously defined items than might be expected.

this replaces the mosquitto module completely. we now have a hierarchical config
that flattens out to the mosquitto format (hopefully) without introducing spooky
action at a distance.

pennae commented:
rewrote the module a little. only very lightly tested and just the first thing that actually ran the test, will probably change quite a few things still.

pennae commented:
okay, this is the general idea now. is this acceptable?

not quite happy with having freeform settings as extra attributes per level instead of having them on the same level as the other options, but putting them on the same level makes validation hard. if there's a way to query all freeform options of a submodule validation would be easy, but haven't found anything of the sort :(

pennae committed:
nixos/mosquitto: rewrite the test

expand the test to check all four forms of passwords, tls certificates (both
server and client), and that acl files are formatted properly.

peterhoeg commented:
You've put in a ton of effort @pennae - I need to take some time out to go through this properly and in detail to give it the attention it deserves. This will realistically only happen this weekend.

pennae commented:
no rush, and no worries. if it's bad it's bad, effort or no :D

pennae commented:
don't want to be pushy, but is there a chance we'll get a review this weekend?

peterhoeg commented:
I plan to take a look this weekend.

pennae commented:
thank you! :+1:

pennae commented:
@peterhoeg heya, it's totally fine if you don't have the time to review! if that is so, is there someone else who could take that over?

peterhoeg commented:
Hey, no, sorry, I haven't. And realistically it isn't going to happen any week night (during the week). It's not cool to leave you hanging when you've put in all this effort. I *will* get to it but yes if anybody else wants to take a stab, please do.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/548

pennae commented:
friendly ping

pennae commented:
i guess the total lack of anything happening is a good sign that we should just abandon this.

tomberek reviewed (changes requested):
Breaking changes would make release notes helpful. Goal would be to merge prior to the next release cycle.

pennae commented:
added release notes, reran tests (all pass).

tomberek reviewed (changes requested):
The home-assistant NixOS tests now do not evaluate.

pennae commented:
rebased, fixed home-assistant test to use the new config format. succeeds here :)
